src/builder.rs

```diff
@@ -27,52 +27,70 @@ use std::io;
 use std::path::Path;
 use std::path::PathBuf;
 use std::str::FromStr;
+use std::sync::Arc;
 
 use async_trait::async_trait;
 use log::debug;
 use tokio::sync::mpsc;
+use tokio::sync::RwLock;
 
+use crate::client::DeviceClient;
+use crate::connection::DeviceConnection;
 use crate::interface::Interface;
 use crate::interfaces::Interfaces;
 use crate::introspection::AddInterfaceError;
 use crate::store::wrapper::StoreWrapper;
 use crate::store::PropertyStore;
 use crate::transport::{Publish, Receive, Register};
-use crate::AstarteDeviceSdk;
-use crate::EventReceiver;
 
 /// Default capacity of the channels
 ///
-/// This constant is the default bounded channel size for *both* the rumqttc AsyncClient and EventLoop
-/// and the internal channel used by the [`AstarteDeviceSdk`] to send events data to the external receiver.
+/// This constant is the default bounded channel size for *both* the rumqttc AsyncClient and
+/// EventLoop and the internal channel used by the [`DeviceClient`] and [`DeviceConnection`] to send
+/// events data to the external receiver and between each component.
 pub const DEFAULT_CHANNEL_SIZE: usize = 50;
 
 /// Declares the conclusive operation of the device builder.
 ///
 /// This trait is already implemented generically for the [`DeviceBuilder`]
 /// and implementing it should be avoided since it has no practical use.
 pub trait DeviceSdkBuild<S, C> {
-    /// Method that consumes the builder and returns a working
-    /// [`AstarteDeviceSdk`] with the specified settings.
-    fn build(self) -> (AstarteDeviceSdk<S, C>, EventReceiver);
+    /// Method that consumes the builder and returns a working [`DeviceClient`] and
+    /// [`DeviceConnection`] with the specified settings.
+    fn build(self) -> (DeviceClient<S>, DeviceConnection<S, C>);
 }
 
 impl<S, C> DeviceSdkBuild<S, C> for DeviceBuilder<S, C>
 where
     S: PropertyStore,
     C: Publish + Receive + Register + Send,
 {
-    fn build(self) -> (AstarteDeviceSdk<S, C>, EventReceiver) {
-        let (tx, rx) = mpsc::channel(self.channel_size);
+    fn build(self) -> (DeviceClient<S>, DeviceConnection<S, C>) {
+        let (tx_connection, rx_client) = flume::bounded(self.channel_size);
```

Comment:
Is it not possible to use `mpsc::channel(self.channel_size);` also here? What kinds of advantages does `flume` introduce?

Reply:
The `mpsc` (multi producer single consumer) channel can be used only by one `Receiver` since it doesn't provide clone trait for the consumer part. While flume provides a `mpmc` (multi producer multi consumer) channel, which in our case the single producer of the connection can send the `DeviceEvent` from Astarte to all the `Client` though this queue/channel. This let us have a single clonable struct, the `Client`, to both send and receive data from Astarte. 

The only thing to note with flume is that an event can be received only once, so it has to be handled by the receiver. Nevertheless, this is still fine since the users of the SDK (e.g. Edgehog) will use only one Client to receive the `DeviceEvent`, but are still able to clone the client freely.

I think this should be better documented. So, if it's fine, I will add my reasoning to the PR

Reply:
The [Tokyo broadcast channel](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html) is also a mpmc channel implementation, but it suffers from the "slow receiver" problem.